### PR TITLE
feat: add options and results support

### PR DESCRIPTION
**Pull request by hal3e**
Closes: https://github.com/FuelLabs/fuels-rs/issues/421, https://github.com/FuelLabs/fuels-rs/issues/420, https://github.com/FuelLabs/fuels-rs/issues/415

This PR adds support for native rust `Options` and `Results` inside the SDK. We can now work with Sway contracts like:

```rust
struct TestStruct {
    option: Option<Address>
}

enum TestEnum {
    EnumOption: Option<Address>
}

abi MyContract {
    fn get_some_u64() -> Option<u64>;
    fn get_some_address() -> Option<Address>;
    fn get_some_struct() -> Option<TestStruct>;
    fn get_some_enum() -> Option<TestEnum>;
    fn get_some_tuple() -> Option<(TestStruct, TestEnum)>;
    fn get_none() -> Option<Address>;
    fn input_primitive(s: Option<u64>) -> bool;
    fn input_struct(s: Option<TestStruct>) -> bool;
    fn input_enum(e: Option<TestEnum>) -> bool;
    fn input_none(none: Option<Address>) -> bool;
}
```
and

```rust
struct TestStruct {
    option: Option<Address>
}

enum TestEnum {
    EnumOption: Option<Address>
}

pub enum TestError {
    NoAddress: str[5],
}

abi MyContract {
    fn get_ok_u64() -> Result<u64, TestError>;
    fn get_ok_address() -> Result<Address, TestError>;
    fn get_ok_struct() -> Result<TestStruct, TestError>;
    fn get_ok_enum() -> Result<TestEnum, TestError>;
    fn get_ok_tuple() -> Result<(TestStruct, TestEnum), TestError>;
    fn get_error() -> Result<Address, TestError>;
    fn input_ok(ok_address: Result<Address, TestError>) -> bool;
    fn input_error(test_error: Result<Address, TestError>) -> bool;
}
```

### CODE REVIEW

**Review by Salka1988 — APPROVED** [review]
LGTM

File: packages/fuels-core/src/lib.rs [hunk 1]
```diff
@@ -367,6 +367,77 @@ impl Tokenizable for fuel_tx::AssetId {
     }
 }
 
+impl<T> Tokenizable for Option<T>
+where
+    T: Parameterize + Tokenizable,
+{
+    fn from_token(token: Token) -> Result<Self, Error> {
```
Comment (digorithm) [inline packages/fuels-core/src/lib.rs:374]: This is such a simple and brilliant way of solving the `abigen` for `Option`/`Result` 🔥

**Review by digorithm — CHANGES_REQUESTED** [review]
This is _so_ good. What a clean approach, really good job! Just left a tiny nit w.r.t. error handling. Feel free to merge after that.

File: packages/fuels-core/src/lib.rs [hunk 2]
```diff
@@ -427,6 +498,29 @@ impl Parameterize for u64 {
     }
 }
 
+impl<T> Parameterize for Option<T>
+where
+    T: Parameterize + Tokenizable,
+{
+    fn param_type() -> ParamType {
+        let types = vec![ParamType::Unit, T::param_type()];
+        let variants = EnumVariants::new(types).expect("should never happen");
```
Comment (segfault-magnet) [inline packages/fuels-core/src/lib.rs:507]: I'd explain in the expect message that since `!types.is_empty()` holds and `EnumVariants::new` fails only if it had no variants, this should never happen.

File: packages/fuels-types/src/lib.rs [hunk 3]
```diff
@@ -17,6 +17,8 @@ pub mod utils;
 // Since it would have the format `struct foo` or `enum bar`, there is a whitespace.
 pub const STRUCT_KEYWORD: &str = "struct ";
```
Comment (segfault-magnet) [inline packages/fuels-types/src/lib.rs:18]: Although these were once used elsewhere, none of the four now have a usage outside the one place inside this file. 

I'd inline these constants to their only place of usage and, in doing so, clarify their meaning and make the code more readable.

File: packages/fuels/tests/harness.rs [hunk 4]
```diff
@@ -3847,3 +3847,195 @@ async fn test_gas_forwarded_defaults_to_tx_limit() -> Result<(), Error> {
...
+    let expected_none = None;
+    let response = contract_instance.input_none(expected_none).call().await?;
+
+    assert!(response.value);
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_rust_result_can_be_decoded() -> Result<(), Error> {
```
Comment (segfault-magnet) [inline packages/fuels/tests/harness.rs:3957]: I find myself joining decoding and encoding into one test case, seeing as these are e2e and it is less of a problem to do so. 

A sway function that accepts the argument validates its content, and returns the same argument but constructed by sway code so as not to have the bytecode just return your original encoded bytes right back at you.

But that is up to you. With the new macro, these are going to get a lot shorter, then we wouldn't pay the penalty of test code size but only of the time spent executing -- which still hasn't grown unwieldy -- so I'm ok if it remains the way it is.

File: packages/fuels-core/src/code_gen/abigen.rs [hunk 5]
```diff
@@ -1,7 +1,9 @@
 use std::collections::HashMap;
 
 use crate::code_gen::bindings::ContractBindings;
-use crate::constants::{ADDRESS_SWAY_NATIVE_TYPE, CONTRACT_ID_SWAY_NATIVE_TYPE};
+use crate::constants::{
+    ADDRESS_NATIVE_TYPE, CONTRACT_ID_NATIVE_TYPE, OPTION_NATIVE_TYPE, RESULT_NATIVE_TYPE,
```
Comment (segfault-magnet) [inline packages/fuels-core/src/code_gen/abigen.rs:5]: I don't find much benefit in constants that are likely to be used in only one place. Inlining them helps with clarity and if the time comes to change them, you only change them at the one place they're used.

File: packages/fuels/tests/harness.rs (hunk 4 above)
Comment (hal3e) [reply packages/fuels/tests/harness.rs:3957]: I have checked the generics examples you did. I like them! However, I would keep these separate as the function names directly describe what is done and the code is a bit _clearer_

File: packages/fuels/tests/test_projects/options/src/main.sw [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        if let Option::Some(u) = input {
+            return u == 36;
+        }
+        false
+    }
+
+    fn input_struct(input: Option<TestStruct>) -> bool {
+        if let Option::Some(s) = input {
+            if let Option::Some(a) = s.option {
+                return a == ~Address::from(ZERO_B256);
```
Comment (iqdecay) [inline packages/fuels/tests/test_projects/options/src/main.sw:67]: I think this test should be with an address different from the 0 address, just as a sanity check.

File: packages/fuels/tests/test_projects/options/src/main.sw [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            }
+        }
+        false
+    }
+
+    fn input_enum(input: Option<TestEnum>) -> bool {
+        if let Option::Some(test_enum) = input {
+            if let TestEnum::EnumOption(option) = test_enum {
+                if let Option::Some(a) = option {
+                    return a == ~Address::from(ZERO_B256);
```
Comment (iqdecay) [inline packages/fuels/tests/test_projects/options/src/main.sw:77]: same remark here

**Review by iqdecay — CHANGES_REQUESTED** [review]
Supeeer clean, I love it! Some questions around naming (as usual with me lol), and also I'm always wary of testing things only with the zero values, which might act as default and thus would recommend changing for non-zero values.

File: packages/fuels-core/src/code_gen/abigen.rs [hunk 8]
```diff
@@ -214,17 +213,26 @@ impl Abigen {
...
 
-    // Checks whether the given type field is a Sway-native type.
+    // Checks whether the given type field is a native type.
     // It's expected to come in as `"struct T"` or `"enum T"`.
-    // `T` is a Sway-native type if it matches exactly one of
-    // the reserved strings, such as "Address" or "ContractId".
-    fn is_sway_native_type(type_field: &str) -> bool {
+    // `T` is a native `high-level language` or Rust type if it matches exactly one of
+    // the reserved strings, such as "Address", "ContractId", "Option" or "Result"
+    pub fn is_native_type(type_field: &str) -> bool {
```
[the remark this answers is not in the document]
Comment (hal3e) [reply packages/fuels-core/src/code_gen/abigen.rs:220]: There is a issue to remove all sway wording in the sdk. So `is_sway_native_type` is not an option IMO. I am open for other suggestions.

File: packages/fuels-types/src/lib.rs [hunk 9]
```diff
@@ -13,11 +13,6 @@ pub mod param_types;
 pub mod parse_param;
 pub mod utils;
 
-// Both those constants are used to determine if a type field represents an `Enum` or a `Struct`.
-// Since it would have the format `struct foo` or `enum bar`, there is a whitespace.
-pub const STRUCT_KEYWORD: &str = "struct ";
-pub const ENUM_KEYWORD: &str = "enum ";
-
```
[the remark this answers is not in the document]
Comment (hal3e) [reply packages/fuels-types/src/lib.rs:20]: I actually agree with @segfault-magnet on this. IMO it is better now. We only use it once and use it exactly where they are needed.

File: packages/fuels/tests/test_projects/options/src/main.sw (hunk 6 above)
Comment (hal3e) [reply packages/fuels/tests/test_projects/options/src/main.sw:67]: We are testing that in other parts of the code but I agree. For sanity let's change it to something :smile:

File: packages/fuels-core/src/code_gen/abigen.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (iqdecay) [reply packages/fuels-core/src/code_gen/abigen.rs:220]: would "fuel-native" types work?

[the remark this answers is not in the document]
Comment (digorithm) [reply packages/fuels-core/src/code_gen/abigen.rs:220]: As @hal3e said, we're removing Sway wording; the SDK is an API to the VM and should be language agnostic. 

In a way, these keywords (option, result, etc) are also native to the VM. So I'm ok with the function being called as it is 🙂

**Review by iqdecay — APPROVED** [review]
Agreed with all answers, let's :ship: it!

**Review by digorithm — APPROVED** [review]
